Ensure proper colorer range for depth range

milenovic opened the pull request:
resolves #161

milenovic committed:
Ensure proper colorer range for depth range

resolves #161

asl commented:
Will you please rebase the PR?

asl commented on graph/assemblygraph.cpp:
```diff
@@ -51,6 +51,7 @@
 
 AssemblyGraph::AssemblyGraph()
         : m_sequencesLoadedFromFasta(NOT_READY)
+        , m_currentScope(graph::Scope::wholeGraph())
```
I do not think this is a proper approach. AssemblyGraph is a data structure and it should not "know" about the drawing scope, so this way you're leaking the abstraction through unrelated data structure.

Certainly, we do have some legacy issues (e.g. global settings state that is changed in different places, but we should improve the situation, and not make it worse).

asl commented on graph/nodecolorer.cpp:
```diff
@@ -76,8 +76,13 @@ QColor DepthNodeColorer::get(const GraphicsItemNode *node) {
     if (g_settings->autoDepthValue) {
-        lowValue = m_graph->m_firstQuartileDepth;
-        highValue = m_graph->m_thirdQuartileDepth;
+        if (m_graph->currentScope().graphScope() == DEPTH_RANGE) {
```
So, instead of fetching the scope from the graph (which is not really something belongs to it), it would be better to route it through `DepthNodeColorer`, or rather calculate low / high value when colorer is created.

asl reviewed (commented):
Thanks for the PR. I think this is a good step forward, though we need to have somehow better abstraction for stateful node colorers. Right now they fetch their state from the global settings which is not something desired.

Please see my comments

milenovic committed:
Reimplement proper colorer range for depth range

When graph is redrawn, scope is passed to colorer which calculates low and high values and stores then for colorers to use when needed. Is this correct way to store them?
resolves #161

asl commented on ui/mainwindow.cpp:
```diff
@@ -1025,6 +1025,7 @@ void MainWindow::drawGraph() {
     resetScene();
     g_assemblyGraph->resetNodes();
     g_assemblyGraph->markNodesToDraw(scope, startingNodes);
+    g_settings->nodeColorer->setGlobalScope(scope,g_assemblyGraph->m_firstQuartileDepth, g_assemblyGraph->m_thirdQuartileDepth);
```
This does not look right. Colorers are transient, they are recreated as necessary, so, on colorer switch this will be lost.

asl commented on graph/nodecolorer.h:
```diff
@@ -51,7 +53,15 @@ class INodeColorer {
 
     [[nodiscard]] NodeColorScheme scheme() const { return m_scheme; }
+
+    static void setGlobalScope(const graph::Scope &scope, double firstQuartileDepth, double thirdQuartileDepth);
```
Why this is a part of common interface? It only belongs to depth colorer.

milenovic commented:
I am not sure I understand how to do it best. Colorers are transient, so that means we need to store either the scope or the calculated lowValue and highValue somewhere so they can be acessed by the depth colorer, correct?
Colorer can be changed without redrawing the graph, so I didn't see a way to access scope from the depth colorer in that case.

milenovic commented:
I tried another implementation where scope is passed only to the DepthNodeColorer, which stores it in a private static variable, which persists when colorers are changed. However, the issue arises if some other colorer is selected when the user selects Scope: Depth range, and then changes the colorer to DepthNodeColorer later without redrawing. Because no instance of DepthNodeColorer existed at the time the graph was drawn, the variable was not saved. It the user presses Draw graph again everything works. 

In my mind, this means that scope needs to be stored at the INodeColorer level. Or just the calculated values as I did in my last commit e48430ce265673e021617ad3a7522a1de05e5656.

Is there a better way than storing _something_ in the common class?

asl commented:
The colorer is switched in a single place: https://github.com/asl/BandageNG/blob/dev/ui/mainwindow.cpp#L1364 (while it is currently in settings, it just it was not untangled fully from the global state, it really should belong to mainwindow, or rather we'd need to have `UI Model` speaking in MVVM terms).

So, you might want to check if the current scheme is coloring by depth, then cast interface to instance of DepthColorer and pass the scope there. The DepthColorer would recalculate min / max values then.

milenovic committed:
Save scope reference in DepthNodeColorer

Store Scope reference when graph is redrawn. Scope reference is stored only in a private static variable of DepthNodeColorer, which persists when colorers are changed.
As DepthNodeColorer instance might not exist when scope is changed, reference is saved by a function defined in INodeColorer.
resolves #161

milenovic committed:
Dynamic implementation of proper range for depth node colorer

resolves #161

milenovic committed:
Fix failed linux tests

fixes #161 build errors

milenovic committed:
Revert "Fix failed linux tests"

This reverts commit 116882a62a4f4845b42eda186d92608b4b15a89e.

milenovic committed:
Install imagemagick on Linux before running tests

Ubuntu >24.04 image does not longer include imagemagick by default.
fixes #161 Linux tests fails

milenovic commented:
Is my latest approach in 91f7cf281d48974d1caa660d734d596e9667ccfd adequate?
Along the way, in c31a7f993ddf52a18cdc2e7da118956669575160 I fixed the dependency issue that caused linux tests to fail as Ubuntu >24.04 image [does not longer include](https://github.com/actions/runner-images/issues/10772#issuecomment-2410774000) imagemagick by default.

asl commented:
Sorry, got distracted to something else. Will check soon!
